usr/lib/common/key.c

```diff
@@ -3185,6 +3185,335 @@ CK_BBOOL kea_priv_check_exportability(CK_ATTRIBUTE_TYPE type)
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
+    }
+
+    found = template_attribute_find(tmpl, CKA_IBM_DILITHIUM_T1, &attr);
+    if (!found) {
+        if (mode == MODE_CREATE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
+    }
+
+    found = template_attribute_find(tmpl, CKA_IBM_DILITHIUM_KEYFORM, &attr);
+    if (!found) {
+        if (mode == MODE_CREATE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
+    }
+
+    return publ_key_check_required_attributes(tmpl, mode);
+}
+
+// dilithium_priv_check_required_attributes()
+//
+CK_RV dilithium_priv_check_required_attributes(TEMPLATE *tmpl, CK_ULONG mode)
+{
+    CK_ATTRIBUTE *attr = NULL;
+    CK_BBOOL found;
+
+    found = template_attribute_find(tmpl, CKA_IBM_DILITHIUM_RHO, &attr);
+    if (!found) {
+        if (mode == MODE_CREATE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
+    }
+
+    found = template_attribute_find(tmpl, CKA_IBM_DILITHIUM_SEED, &attr);
+    if (!found) {
+        if (mode == MODE_CREATE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
+    }
+
+    found = template_attribute_find(tmpl, CKA_IBM_DILITHIUM_TR, &attr);
+    if (!found) {
+        if (mode == MODE_CREATE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
+    }
+
+    found = template_attribute_find(tmpl, CKA_IBM_DILITHIUM_S1, &attr);
+    if (!found) {
+        if (mode == MODE_CREATE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
+    }
+
+    found = template_attribute_find(tmpl, CKA_IBM_DILITHIUM_S2, &attr);
+    if (!found) {
+        if (mode == MODE_CREATE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
+    }
+
+    found = template_attribute_find(tmpl, CKA_IBM_DILITHIUM_T0, &attr);
+    if (!found) {
+        if (mode == MODE_CREATE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
+    }
+
+    found = template_attribute_find(tmpl, CKA_IBM_DILITHIUM_T1, &attr);
+    if (!found) {
+        if (mode == MODE_CREATE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
+    }
+
+    found = template_attribute_find(tmpl, CKA_IBM_DILITHIUM_KEYFORM, &attr);
+    if (!found) {
+        if (mode == MODE_CREATE) {
+            TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+            return CKR_TEMPLATE_INCOMPLETE;
+        }
+    }
```

Comment:
I think the above functions would be more concise if they just implemented a loop like
```
CK_ULONG req_attrs[] = {....}, i;
for (i = 0; i < sizeof(req_attrs) / sizeof(req_attrs[0]; i++) {
  if (template_attribute_find(tmpl, CKA_IBM_DILITHIUM_KEYFORM, &attr)! && mode == MODE_CREATE) {
    TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
    return CKR_TEMPLATE_INCOMPLETE;
  }
}
```

Reply:
Yes, that's true. But as all other *_priv_check_required_attributes and *_publ_check_required_attributes routines are coded in the same way I would leave it as is for now. Let's then re-code all affected routines in a separate "housekeeping" pull-request.

Reply:
Id prefer to have it done now instead postponing it until the day the cows come home (I mean IF the additional effort is not too high).

Reply:
ok, just pushed the changes. I also changed ibm_dilithium_publ_check_required_attributes() to have at least these two routines with the same logic. The cows may come home then ;-)